Allow annotation target override on gateway

BadLiveware opened the pull request:
<!--
    Please read https://github.com/kubernetes-sigs/external-dns#contributing before submitting
    your pull request. Please fill in each section below to help us better prioritize your pull request. Thanks!
-->

**Description**
This allows the `external-dns.alpha.kubernetes.io/target` annotation to work for gateway api resources

<!-- Please provide a summary of the change here. -->

<!-- Please link to all GitHub issue that this pull request implements(i.e. Fixes #123) -->
Fixes #ISSUE

**Checklist**

- [X] Unit tests updated
- [X] End user documentation updated

johngmyers commented:
/lgtm
/assign njuettner

johngmyers commented:
@BadLiveware could you please update `doc/annotations/annotations.md`? Please add a footnote that the annotation should be on the `Gateway`.

BadLiveware commented:
> @BadLiveware could you please update `doc/annotations/annotations.md`? Please add a footnote that the annotation should be on the `Gateway`.

Rebased and updated

johngmyers commented on source/gateway.go:
```diff
@@ -352,8 +352,12 @@ func (c *gatewayRouteResolver) resolve(rt gatewayRoute) (map[string]endpoint.Tar
-					hostTargets[host] = append(hostTargets[host], addr.Value)
+				override := getTargetsFromTargetAnnotation(gw.gateway.Annotations)
+				hostTargets[host] = append(hostTargets[host], override...)
+				if override == nil {
```
```suggestion
				if len(override) == 0 {
```

johngmyers commented on source/gateway_httproute_test.go:
```diff
@@ -1033,6 +1033,128 @@ func TestGatewayHTTPRouteSourceEndpoints(t *testing.T) {
+			},
+		},
+		{
+			title: "AnnotationOverrideMultipleStatusAddresses",
```
This test isn't necessary.

johngmyers commented:
/ok-to-test
/lgtm

BadLiveware commented:
@johngmyers is this waiting on a review by @njuettner ?

johngmyers commented:
@BadLiveware yes. The approvers aren't terribly responsive at the best of times and it's European vacation season.

BadLiveware commented:
Cool, just wanted to make sure, on vacation myself so fully understand :)

On Tue, Jul 25, 2023, 02:58 John Gardiner Myers ***@***.***>
wrote:

> @BadLiveware <https://github.com/BadLiveware> yes. The approvers aren't
> terribly responsive at the best of times and it's European vacation season.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-sigs/external-dns/pull/3452#issuecomment-1648814030>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOFHEYEINPWEVMOKRJ2UFTXR4K43ANCNFSM6AAAAAAVQHSEQM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

johngmyers commented:
/approve

johngmyers commented:
/test pull-external-dns-lint
/test pull-external-dns-unit-test
